File: Ray Tracing next Week/util/week2/motion_blur/sphere.py
import math

from util.week1.vec3 import dot
from util.week1.hittable import Hittable
from util.week2.ray import Ray


class Sphere(Hittable):

    def __init__(
        self,
        center1,
        radius,
        material,
        center2=None
    ):
        """
        ============================================
        SPHERE CONSTRUCTOR
        ============================================

        Supports:

        1. Static sphere
        2. Moving sphere

        --------------------------------------------
        STATIC:

        Sphere(center, radius, material)

        center(t) = center

        --------------------------------------------
        MOVING:

        Sphere(center1, radius, material, center2)

        center(t)=center1+t(center2-center1)

        We store motion as a ray.
        """

        self.radius = max(0.0, radius)
        self.mat = material

        # ============================================
        # STATIC SPHERE
        # OLD behavior
        # ============================================
        if center2 is None:

            # CHANGED:
            # store zero-velocity ray
            self.center = Ray(
                center1,
                center1 * 0
            )

        # ============================================
        # MOVING SPHERE
        # NEW behavior
        # ============================================
        else:

            # center(t)=center1+t(center2-center1)

            self.center = Ray(
                center1,
                center2 - center1
            )

    # ==================================================
    # HIT TEST
    # ==================================================
    def hit(self, r, ray_t, rec):

        """
        ============================================
        STEP 1:
        Compute current center at ray time
        ============================================

        Formula:

        C(t)=C1+t(C2-C1)

        CHANGED:
        Before:
            fixed center

        Now:
            moving center
        """

        current_center = self.center.at(r.time())

        """
        Sphere equation:

        (P-C(t))·(P-C(t)) = r²

        Ray equation:

        P = O+tD
        """

        # ============================================
        # Vector from ray origin to moving center
        # ============================================
        oc = current_center - r.origin()

        # a = D·D
        a = dot(
            r.direction(),
            r.direction()
        )

        # h = D·oc
        h = dot(
            r.direction(),
            oc
        )

        # c = oc·oc - r²
        c = dot(
            oc,
            oc
        ) - self.radius * self.radius

        # ============================================
        # Discriminant
        #
        # h²-ac
        # ============================================
        discriminant = h * h - a * c

        if discriminant < 0:
            return False

        sqrtd = math.sqrt(discriminant)

        # ============================================
        # Find nearest root
        # ============================================
        root = (h - sqrtd) / a

        if not ray_t.surrounds(root):

            root = (h + sqrtd) / a

            if not ray_t.surrounds(root):
                return False

        # ============================================
        # Store hit
        # ============================================
        rec.t = root

        # P=O+tD
        rec.p = r.at(root)

        """
        ============================================
        CHANGED:
        normal uses current_center

        Before:
            rec.p - self.center

        Wrong for moving sphere

        Now:
            rec.p - current_center
        ============================================
        """

        outward_normal = (
            rec.p - current_center
        ) / self.radius

        rec.set_face_normal(
            r,
            outward_normal
        )

        # material
        rec.mat = self.mat

        return True

File: Ray Tracing next Week - Javascript/Ch7-A/sphere.js
class Sphere extends Hittable{
   
  
   constructor({
  center1,
  center2,
  radius,
  material
}) {
  super();

   var rVec = new Vec3(radius, radius, radius)


  if(center2==null){

    this.center = new Ray({origin:center1,direction:Vec3.sub(center1, new Vec3(0,0,0))});
    this.bbox = new AABB({a:Vec3.sub(center1,rVec),b:Vec3.add(center1,rVec)})
    
  }
  else{
    this.center = new Ray({origin:center1,direction:Vec3.sub(center2, center1)});
    var box1 = new AABB({a:Vec3.sub(this.center.at(0), rVec),b:Vec3.add(this.center.at(0), rVec)})
    var box2 = new AABB({a:Vec3.sub(this.center.at(1), rVec),b:Vec3.add(this.center.at(1), rVec)})
    this.bbox = new AABB({box0:box1,box1:box2})
  }

  this.radius = Math.max(0, radius);
  this.material = material;
} 


   bounding_box(){
    return this.bbox
   }


  get_sphere_uv(p,u,v) {
        // p: a given point on the sphere of radius one, centered at the origin.
        // u: returned value [0,1] of angle around the Y axis from X=-1.
        // v: returned value [0,1] of angle from Y=-1 to Y=+1.
        //     <1 0 0> yields <0.50 0.50>       <-1  0  0> yields <0.00 0.50>
        //     <0 1 0> yields <0.50 1.00>       < 0 -1  0> yields <0.50 0.00>
        //     <0 0 1> yields <0.25 0.50>       < 0  0 -1> yields <0.75 0.50>

        var theta = Math.acos(-p.y);
        var phi = Math.atan2(-p.z, p.x) + Math.PI;

        u(phi / (2*+ Math.PI));
        v(theta / + Math.PI);
    }


   hit(r,ray_t,hit_record){
    //super(r,ray_tmin,ray_tmax,rec)

       // var current_center = this.center.at(r.time())
        var current_center = this.center.origin()

        var oc = Vec3.sub(current_center,r.origin())
        //console.log(r.direction())
        var a = r.direction().length_squared();
        var h = Vec3.dot(r.direction(), oc);
        var c = oc.length_squared() - this.radius*this.radius;
        var discriminant = h*h - a*c;
    
        if (discriminant < 0) {
            return false;
        } 
        
        var sqrtd = Math.sqrt(discriminant);


        var root = (h - sqrtd) / a;
      
        /*
        if (root <= ray_tmin || ray_tmax <= root) {
            root = (h + sqrtd) / a;
            if (root <= ray_tmin || ray_tmax <= root)
                return false;
        }
            */

        if (!ray_t.surrounds(root)) {
            root = (h + sqrtd) / a;
            if (!ray_t.surrounds(root))
                return false;
        }



        var rec = new HitRecord()
        rec.t = root 
        rec.p = r.at(rec.t);
        var outward_normal = Vec3.div(Vec3.sub(rec.p,current_center),this.radius)
        rec.set_face_normal(r,outward_normal)
        this.get_sphere_uv(outward_normal,u=>{rec.u=u},v=>{rec.v=v})
        rec.material = this.material
        hit_record(rec)
        return true
   }

}